### DISCUSSION TITLE
Problems: Extracting the complete Method Call Hierarchy of a complicated Java method.

### DISCUSSION THREAD

**Discussion by oriana19993926782** [github/codeql#13951] [Q&A] [answered]
**I'm attempting to prompt chatGPT to generate unit tests for a method within the [apache/shardingsphere-elasticjob](https://github.com/apache/shardingsphere-elasticjob)**  project. 

For this purpose, I need to 
1. Extract the body of the target method, which I've termed "Method A".
2. Get the complete call hierarchy or call graph of "Method A", as it might invoke another method, "Method B", which could further call "Method C", and so on. This is crucial for chatGPT to understand the exact behavior of "Method A". (Method A B C here are self-defined methods, not from some third-party library)

The body of my target method _setUpJobConfiguration(String, JobConfiguration)_ is down below, where you can see it further calls
**Non-static** methods:   _checkConflict Job(String, JobConfiguration)_ (for example)
**Static** methods:  _YamlEngine.marshal_,  _JobConfigurationPOJO.fromJobConfiguration_
```
public JobConfiguration setUpJobConfiguration(final String jobClassName, final JobConfiguration jobConfig) {
    checkConflictJob(jobClassName, jobConfig);
    if (!jobNodeStorage.isJobNodeExisted(ConfigurationNode.ROOT) || jobConfig.isOverwrite()) {
        jobNodeStorage.replaceJobNode(ConfigurationNode.ROOT, YamlEngine.marshal(JobConfigurationPOJO.fromJobConfiguration(jobConfig)));
        jobNodeStorage.replaceJobRootNode(jobClassName);
        return jobConfig;
    }
    return load(false);
}
```

**Two big problems** I ran into that cause me only getting an incomplete call graph:

1. **Unable to capture the specific overridden implementation method body of almost every interface method invoked within this method body, directly or indirectly. This also means I'm unable to further navigate the call graph within these implementation bodies.**

Example Interface method:`void persist(String key, String value);`

Example overridden implementation on this interface method:
```
  @Override
    public void persist(final String key, final String value) {
        try {
            if (!isExisted(key)) {
                client.create().creatingParentsIfNeeded().withMode(CreateMode.PERSISTENT).forPath(key, value.getBytes(StandardCharsets.UTF_8));
            } else {update(key, value);}
        //CHECKSTYLE:OFF
        } catch (final Exception ex) {
        //CHECKSTYLE:ON
            RegExceptionHandler.handleException(ex);
        }}
```

2. **Unable to get the body of self-defined static methods**, such as_YamlEngine.marshal_, _JobConfigurationPOJO.fromJobConfiguration_. (Note that when I deleted 'and overriddenMethod.fromSource()' and 'overriddenMethod.getBody(),' from my query, I can see these two static methods in results, but I still cannot get the body of these two static methods, which are not from some third-party compiled library.)

Below is the codeQL query I wrote to achieve my purpose. I'm currently stuck here and **greatly appreciate guidance on these challenges and any enhancements that can be made to my query:**

```
import java

// Predicate to determine if a 'caller' either directly or indirectly calls another 'callee'.
predicate callsRecursively(Callable caller, Callable callee) {
  caller.polyCalls(callee)

  or 
  // Checks if 'caller' calls an interface method and 'callee' is a potential target of that call.
  exists(Callable interfaceMethod, MethodAccess ma |
    caller.polyCalls(interfaceMethod) and
    ma.getMethod() = interfaceMethod and
    ma.getAChildExpr*().(MethodAccess).getMethod() = callee 
  )
  
  or
  // Recursively checks if 'caller' indirectly calls 'callee' through some other method 'mid'.
  exists(Callable mid|
    caller.polyCalls(mid) and
    callsRecursively(mid, callee)
  )
}

predicate hasMethodCall(Expr expr) {
  expr instanceof MethodAccess
  or
  exists(Expr child |
    child = expr.getAChildExpr() and
    hasMethodCall(child))
}


class TargetMethod extends Method {
  TargetMethod() {
    this.getName() = "setUpJobConfiguration" 
  }
}

from TargetMethod tm, Method calledMethod, Method overriddenMethod
where
callsRecursively(tm, calledMethod) and
// Check if 'overriddenMethod' either overrides 'calledMethod' or is the same as 'calledMethod'
// and ensure that 'overriddenMethod' is not abstract
(overriddenMethod.overrides*(calledMethod) and not overriddenMethod.isAbstract() or overriddenMethod = calledMethod) 
// Ensure that 'overriddenMethod' are from source code and not from compiled libraries
and overriddenMethod.fromSource()
  
select
  tm.getSignature(), 
  tm.getBody(), 
  "This is a target method",
  calledMethod,
  overriddenMethod,
  overriddenMethod.getBody(),
  "This method is called directly or indirectly in the target method."
```

**Comment by aibaars** [comment]
I tried running the query on `apache/shardingshere`  but got no results because that code base does not contain a method named `setUpJobConfiguration`.

The missing body of `YamlEngine.marshal` is strange; that class exists in the repository. However, it seems to be missing somehow from the CodeQL database. Instead CodeQL reports `/home/runner/work/shardingsphere/shardingsphere/infra/util/target/classes/org/apache/shardingsphere/infra/util/yaml/YamlEngine.class` as the path of the `YamlEngine` class. Class files do not contain the source code of the method body, so that explains why `getBody` does not return anything. It looks like CodeQL did not "see" the `YamlEngine.java` source file. This typically happens when a source tree contains build artefacts from a previous compilation run. In such cases the Java compiler may re-use existing class files and skip the source files. As a result CodeQL won't analyse the source file. In case of apache/shardingsphere, I don't understand how this could happen. The project seems to run [`mvn clean install`](https://github.com/apache/shardingsphere/blob/1fc51da01acd39d113dfd2f1db2678f941cbe6ba/.github/workflows/nightly-check.yml#L167-L168) which should get rid of any `.class` files.

Looking at the [documentation](https://codeql.github.com/docs/codeql-language-guides/navigating-the-call-graph/) it looks like `polyCalls` already includes methods that override other methods. I think the `Method overriddenMethod` should not be needed. You can probably replace `caller.callsRecursively(callee)` with ` caller.polyCalls+(callee)`.

I think the following block is not doing what you need; it finds cases where a `caller` calls some other method and then finds all accesses to that other method anywhere in the code. 
```ql
  // Checks if 'caller' calls an interface method and 'callee' is a potential target of that call.
  exists(Callable interfaceMethod, MethodAccess ma |
    caller.polyCalls(interfaceMethod) and
    ma.getMethod() = interfaceMethod and
    ma.getAChildExpr*().(MethodAccess).getMethod() = callee 
  )
```

**Reply by oriana19993926782** [reply]
Really really sorry, I've made a stupid mistake. It should be [apache/shardingsphere-elasticjob](https://github.com/apache/shardingsphere-elasticjob) instead of [apache/shardingsphere](https://github.com/apache/shardingsphere). That's why you got 0 results running my query. I have already corrected this mistake in my question.

**Comment by aibaars** [comment]
I think the problem is in the CodeQL database . Some source files are missing and methods resolve to class files instead. For example the definition of the  `persist` method is found in `RegistryCenter.java`, but the implementations are found only in class files `ZookeeperRegistryCenter.class` and `ZookeeperRegistryCenterTestUtil.class`
```
from Method m, string f
where m.getName() = "persist" and f = m.getFile().getRelativePath()
select m, f
```

| # | m | f
| -- | -- | --
| 1 | persist | elasticjob-infra/elasticjob-registry-center/elasticjob-registry-center-api/src/main/java/org/apache/shardingsphere/elasticjob/reg/base/RegistryCenter.java |
| 2 | persist | elasticjob-infra/elasticjob-registry-center/elasticjob-regitry-center-provider/elasticjob-registry-center-zookeeper-curator/target/classes/org/apache/shardingsphere/elasticjob/reg/zookeeper/ZookeeperRegistryCenter.class |
| 3 | persist | elasticjob-infra/elasticjob-registry-center/elasticjob-regitry-center-provider/elasticjob-registry-center-zookeeper-curator/target/test-classes/org/apache/shardingsphere/elasticjob/reg/zookeeper/util/ZookeeperRegistryCenterTestUtil.class |

The same thing happens for the `marshal` method.

| # | m | f |
| -- | -- | -- |
| 1 | marshal | elasticjob-infra/elasticjob-infra-common/target/classes/org/apache/shardingsphere/elasticjob/infra/yaml/YamlEngine.class |


I have no idea why the database is incomplete, but it explains why you aren't able to get the bodies of those methods. We'll need to investigate this further.

**Reply by aibaars** [reply, marked as answer]
The cause of the incomplete database is that CodeQL did not support Lombok and skips any files that import lombok annotations. Support for Lombok is experimental and can be enabled by setting the environment variable `CODEQL_EXTRACTOR_JAVA_RUN_ANNOTATION_PROCESSORS`. You can create a complete database with the CodeQL CLI as follows:

```
CODEQL_EXTRACTOR_JAVA_RUN_ANNOTATION_PROCESSORS=true codeql database create -ljava -s shardingsphere-elasticjob shardingsphere-elasticjob-database
```

**Reply by oriana19993926782** [reply]
**Thank you so so so so much💗. This is very very helpful!!!!!!** I have used CodeQl CLI to create the database, run my query, and have seen the implementations with methods invoked within and static methods as well. 

Also, I took your advice that I could just use `polyCalls+`. I feel a bit silly for having gone the long way and writing a `callsRecursively` predicate 😂. My query looks much neater now. But it is a little strange that even though this `polyCalls+` works like magic, there doesn't seem to be any documentation on it, only on `polyCalls`.

On another note, is there a way to enable this` CODEQL_EXTRACTOR_JAVA_RUN_ANNOTATION_PROCESSORS=true` feature if I want to add this database from Github in VS Code instead? Or does it has to be created using CodeQL CLI? Also, I could not find documentation regarding this` CODEQL_EXTRACTOR_JAVA_RUN_ANNOTATION_PROCESSORS` neither. If you come across it, could you kindly share the link with me?

**Reply by aibaars** [reply]
The `+` in `polyCalls+` is a QL feature that means apply `polyCalls` 1 or more times. There is also `*` which means zero or more times. You can use `*` and `+` on other predicates too. It's a short way of writing a simple recursion.

`CODEQL_EXTRACTOR_JAVA_RUN_ANNOTATION_PROCESSORS` is an undocumented feature flag. It will probably be removed once Lombok support is no longer experimental. Most databases downloaded from GitHub via VSCode will likely work with Lombok in a few weeks time.

**Reply by aibaars** [reply]
@oriana19993926782 The database for [apache/shardingsphere-elasticjob](https://github.com/apache/shardingsphere-elasticjob) has been updated and should contain the previously missing Java sources files now.
